akka-docs/src/main/paradox/scala/guide/tutorial_3.md

```diff
@@ -1,253 +1,184 @@
+:   @@snip [DeviceInProgress.scala]($code$/scala/tutorial_3/DeviceInProgress.scala) { #device-with-read }
 
 Java
-:   @@snip [DeviceGroupTest.java]($code$/java/jdocs/tutorial_3/DeviceGroupTest.java) { #device-group-test3 }
-
-So far, we have implemented everything for registering device actors in the group. Devices come and go, however, so
-we will need a way to remove those from the @scala[`Map[String, ActorRef]`] @java[`Map<String, ActorRef>`]. We will assume that when a device is removed, its corresponding device actor
-is simply stopped. We need some way for the parent to be notified when one of the device actors are stopped. Unfortunately,
-supervision will not help because it is used for error scenarios, not graceful stopping.
-
-There is a feature in Akka that is exactly what we need here. It is possible for an actor to _watch_ another actor
-and be notified if the other actor is stopped. This feature is called _Death Watch_ and it is an important tool for
-any Akka application. Unlike supervision, watching is not limited to parent-child relationships, any actor can watch
-any other actor given its `ActorRef`. After a watched actor stops, the watcher receives a `Terminated(ref)` message
-which also contains the reference to the watched actor. The watcher can either handle this message explicitly or, if
-it does not handle it directly it will fail with a `DeathPactException`. This latter is useful if the actor cannot
-longer perform its duties after its collaborator actor has been stopped. In our case, the group should still function
-after one device have been stopped, so we need to handle this message. The steps we need to follow are the following:
-
- 1. Whenever we create a new device actor, we must also watch it.
- 2. When we are notified that a device actor has been stopped we also need to remove it from the @scala[`Map[String, ActorRef]`] @java[`Map<String, ActorRef>`] which maps
-    devices to device actors.
-
-Unfortunately, the `Terminated` message only contains the `ActorRef` of the child actor but we do not know
-its ID, which we need to remove it from the map of existing device to device actor mappings. To be able to do this removal, we
-need to introduce another placeholder, @scala[`Map[ActorRef, String]`] @java[`Map<ActorRef, String>`], that allow us to find out the device ID corresponding to a given `ActorRef`. Putting
-this together the result is:
+:   @@snip [DeviceInProgress2.java]($code$/java/jdocs/tutorial_3/inprogress2/DeviceInProgress2.java) { #device-with-read }
 
-Scala
-:   @@snip [DeviceGroup.scala]($code$/scala/tutorial_3/DeviceGroup.scala) { #device-group-remove }
+Note in the code that:
 
-Java
-:   @@snip [DeviceGroup.java]($code$/java/jdocs/tutorial_3/DeviceGroup.java) { #device-group-remove }
+* The helper object defines how to construct a `Device` actor. The `props` parameters include an ID for the device and the group to which it belongs, which we will use later.
```

Comment:
There is no helper object in Java, there it is a static method, and in Scala it is called "companion object"